### ISSUE TITLE
Current working directory: `cwd` instead of `pwd`?

### ISSUE THREAD

**Issue by shardulc** [com-lihaoyi/os-lib#215]
I’m relatively new to Scala and just started using this library, so sorry if I’m just unaware of preexisting conventions! But it seems to me like `os.pwd` should rather be named `os.cwd` for the following reasons (mostly about consistent nomenclature):

 * The docstring is “The **c**urrent **w**orking **d**irectory for this process.” (emphasis added) ([source](https://github.com/com-lihaoyi/os-lib/blob/e741e2acfe590fa8c72dd0adba3d07118e35650d/os/src-jvm/package.scala#L25))
 * The `pwd` Unix utility is described as “**print** name of current/working directory” (emphasis added) ([source](https://linux.die.net/man/1/pwd)), whereas the `cwd` module and the `getcwd` function therein are for just getting its pathname in a program ([source](https://linux.die.net/man/3/cwd))

To be fair, there is a POSIX-defined environment variable called `PWD` that the utility `pwd` reads in some cases ([Unix SE answer](https://unix.stackexchange.com/questions/79571/symbolic-link-recursion-what-makes-it-reset/79621#79621) with more details), but I’m not sure if this `os.pwd` ever reaches that. It is currently implemented by calling out to [`java.nio.file.Path#toAbsolutePath`](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Path.html#toAbsolutePath--) which uses the JVM system property `user.dir` which, in all the [documentation](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#getProperties--) I was able to find, is only described as the user’s current working directory, and I don’t know how it is set.

Was there some other reason this was named `pwd` instead of `cwd`?

Are there existing norms / rules for this library for when and how to deprecate an API?

**Comment by lefou** [comment]
I agree with you that `cwd` would be the better name. I also stumbled upon this myself. Yet, we try to provide a stable API and `pwd` isn't so much off. It can be found in the API and it's well documented. Breaking the API just to be a bit more correct is still breaking the API, which we really try to avoid in a small but universal utility library like this.

We could introduce a `cwd` and deprecate the `pwd`, but besides producing one more warning, it has potential to irritate the users too. Also, we don't plan a breaking release for the near future, so both names would stay most likely for a long time, which isn't that ideal either.

I'm not against that change per se, but I think it isn't worth it.

**Comment by shardulc** [comment]
Thanks for your attention, and fair enough! Everything is a tradeoff. To be clear, I indeed meant to suggest a deprecation rather than a breaking API change. But mostly just wanted to make sure the developer(s) were aware of this.

This issue can be closed if needed.

**Comment by SethTisue** [comment]
> We could introduce a cwd and deprecate the pwd

fwiw, this is what I would suggest

**Comment by lefou** [comment]
@lihaoyi @lolgab Any opinions on that?

**Comment by lihaoyi** [comment]
I don't have a strong opinion here

**Comment by jodersky** [comment]
I am against changing it, for the practical reasons mentioned.

The issue is that this is a very core and public definition of os-lib, that is likely called in many many code bases. Also, I don't think that `os.pwd` is *that* confusing, so IMO the gains here are not warranted by the annoyances this would cause.

It's unfortunate, but I would say that the ship has sailed on that one.

**Comment by sake92** [comment]
It's an interesting rabbithole to get into.. 😄  

This SO answer says that they are mostly interchangeable: https://unix.stackexchange.com/a/709897
There is the `$PWD` env var in most *nix shells that returns.. current working directory.

---
In the Notes section here https://linux.die.net/man/3/cwd#Notes
they mention this paragraph:

> Actually, on Mac OS , the "getcwd()", "fastgetcwd()" and "fastcwd()" functions are all aliases for the "cwd()" function, which, **on Mac OS , calls 'pwd**'. 

---

TLDR:
The `pwd` could also be read as "present working directory", which also makes sense.  
Most people have probably used that command and are familiar what it means.  
So I'd keep the current state of things.  


---
A side note, to me the `cwd` used in this context is more confusing https://github.com/com-lihaoyi/os-lib/blob/94b26916b95933fd4fe1621a1108d6a8622f91b5/os/src-jvm/ProcessOps.scala#L41  
I'd call it just `wd` (working directory). Because it is *current* for the process I want to spawn, not current-current (for the current process).. 😅

**Comment by lefou** [comment]
Thank you all for your feedback! We'll keep `pwd`.
